### PR TITLE
Ip tests update

### PR DESCRIPTION
**Pull request by scottnath**
Upgrade to input plugin tests to allow for radio, select, checkboxes, and textareas.

**Changelog**

```markdown
**New**

* Input plugin fixtures added
* Tests options, if they exist

**Changed**

* Testing upgraded to check multi-inputs like radios/checkboxes
```
Resolves: #11 
---

`DCO 1.1 Signed-off-by: Scott Nath <github@scottnath.com>`

### CODE REVIEW

**Comment by scottnath** [comment]
@Snugug @rachelnicole please look over this PR

**Comment by Snugug** [comment]
I'm thinking that maybe `content-types` isn't the right place for these tests to live? We update this module fairly frequently and the input plugins don't actually rely on it outside of this test file. It'd make greenkeeper happier if we split this out in to another module that they could rely on

**Comment by Snugug** [comment]
Thoughts?

File: lib/input-plugin-tests.js [hunk 1]
```diff
@@ -73,45 +71,57 @@ module.exports = function pluginTests(test, plugin) {
       t.true(imp.hasOwnProperty('type'), 'Input has a `type` property');
       t.true(typeof imp.type === 'string', 'Input type is a string');
 
+      if (imp.hasOwnProperty('options')) {
+        t.true(Array.isArray(imp.options), 'Options is an object');
```
Comment (Snugug) [inline lib/input-plugin-tests.js]: If `Array.isArray(imp.options)` is true, then it is _not_ an object, but an array. Should it be an object or an array

File: package.json [hunk 2]
```diff
@@ -50,6 +50,7 @@
     "input-plugin-quote": "0.0.1",
     "input-plugin-text": "^0.0.5",
     "input-plugin-textarea": "^0.0.1",
+    "node-dir": "^0.1.12",
```
Comment (Snugug) [inline package.json]: why was this dependency added? Don't see it used anywhere

**Comment by scottnath** [comment]
@Snugug my thought on your thought about thinking about moving this stuff outta here is that your think is a good think. Can we leave it in here for now and I'll pull it out next week? (I'm ooto today, obviously) That way @rachelnicole can complete her stories for this sprint.
Name of new fancy module?

File: package.json (hunk 2 above)
Comment (scottnath) [inline package.json]: Not _added_, just an accidental upgrade. It's used by load.js.

File: lib/input-plugin-tests.js (hunk 1 above)
Comment (scottnath) [inline lib/input-plugin-tests.js]: fixed

**Comment by scottnath** [comment]
@Snugug changes made. merges merged. [stories written](https://github.com/punchcard-cms/content-types/issues/36)

File: package.json (hunk 2 above)
Comment (Snugug) [inline package.json]: then why is it in our direct dependency stack?

**Comment by Snugug** [comment]
@scottnath I haven't done a full review of this, those two things were first to catch my eye. Will continue with review probably Monday given my scheule

**Comment by scottnath** [comment]
No worries.
@rachelnicole you can use my branch for this pr to complete your work on
your input plugins.

**Comment by rachelnicole** [comment]
@scottnath awesome. i'm going to work on the image plugin more today, but if this isn't merged in on Monday I'll use the branch for my work.

**Comment by Snugug** [comment]
Version number was removed with the move to automated releases.

**Comment by rachelnicole** [comment]
@Snugug is there a different way I can include it then via `git+ssh//git@github.com:scottnath/content-types.git#ip-tests-update` so I can work on my stuff that requires the new changes? because I can't `npm install` it that way without one, it just tells me no version provided.

**Comment by scottnath** [comment]
@Snugug agreed on @rachelnicole 's issue needing a fix. Is there a way to have the version number in there and instead be overwritten during automated release? We need to be able to access branches and/or alternate releases.

File: package.json (hunk 2 above)
Comment (scottnath) [inline package.json]: This is in the direct deps stack because it is used by [load.js](https://github.com/punchcard-cms/content-types/blob/master/lib/content-types/load.js) which is not a development file, it is part of the main system

**Comment by scottnath** [comment]
@Snugug I have found these instructions:
https://gist.github.com/boennemann/54042374e49c7ade8910
from `semantic-release`'s owner. Looks pretty straightforward...however, we should have some kind of plan for this. Should we have multiple `beta` or `dev` tags? 

Should I attempt on my fork? Would that even work or does it have to happen on the main (punchcard-cms) user's account?

**Comment by scottnath** [comment]
@rachelnicole 
cc: @Snugug 

I have created a new branch here:
https://github.com/scottnath/content-types/tree/ip-tests-update-version
which is exactly the same as this PR, but with a version number (0.8.0) in `package.json`.

@rachelnicole please use this branch to complete your input plugin work.

File: package.json (hunk 2 above)
Comment (Snugug) [inline package.json]: ok

File: tests/fixtures/input-plugins/input-plugin-checkbox.js [hunk 3]
```diff
@@ -33,5 +33,5 @@ module.exports = {
       },
     },
   },
-  html: `<label for="{{checkboxSelector.id}}">{{checkboxSelector.label}}</label>{% for option in checkboxSelector.options %}<label><input type="{{checkboxSelector.type}}" name={{checkboxSelector.name}} id="{{checkboxSelector.id}}" value="{{checkboxSelector.value}}" {% if option.value == checkboxSelector.value %}checked{% endif %}>{{option.label}}</label>{% endfor %}`,
-};
\ No newline at end of file
+  html: `{% for option in checkboxSelector.options %}<label for="{{checkboxSelector.id}}--{{option.index}}"><input type="{{checkboxSelector.type}}" name="{{checkboxSelector.name}}" id="{{checkboxSelector.id}}--{{option.index}}" value="{{option.value}}" {% if option.value == checkboxSelector.value %}checked{% endif %}>{{option.label}}</label>{% endfor %}`,
```
Comment (Snugug) [inline tests/fixtures/input-plugins/input-plugin-checkbox.js]: With checkboxes, it's possible for multiple checkboxes to be selected. Should take that in to account.

Comment (Snugug) [inline tests/fixtures/input-plugins/input-plugin-checkbox.js]: If we're going to have multiple checkboxes, we need to signal to the form generation that this should be wrapped in a fieldset

File: tests/fixtures/input-plugins/input-plugin-multiple-textarea.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        on: 'blur',
+      },
+      label: 'Snarky Rant',
+      type: 'textarea',
+      settings: {
+        empty: false,
+      },
+    },
+  },
+  html: '<label for="{{codeSnippet.id}}">{{codeSnippet.label}}<textarea type="{{codeSnippet.type}}" id="{{codeSnippet.id}}" name="{{codeSnippet.name}}">{{codeSnippet.value}}</textarea></label><label for="{{snarkyRant.id}}">{{snarkyRant.label}}<textarea type="{{snarkyRant.type}}" id="{{snarkyRant.id}}" name="{{snarkyRant.name}}">{{snarkyRant.value}}</textarea></label>',
```
Comment (Snugug) [inline tests/fixtures/input-plugins/input-plugin-multiple-textarea.js]: Can we _not_ have snarky rant as an example?

File: tests/fixtures/input-plugins/input-plugin-radio.js [hunk 5]
```diff
@@ -33,5 +33,5 @@ module.exports = {
       },
     },
   },
-  html: `<label for="{{radioSelector.id}}">{{radioSelector.label}}</label>{% for option in radioSelector.options %}<label><input type="{{radioSelector.type}}" name={{radioSelector.name}} id="{{radioSelector.id}}" value="{{radioSelector.value}}" {% if option.value == radioSelector.value %}checked{% endif %}>{{option.label}}</label>{% endfor %}`,
-};
\ No newline at end of file
+  html: `{% for option in radioSelector.options %}<label for="{{radioSelector.id}}--{{option.index}}"><input type="{{radioSelector.type}}" name="{{radioSelector.name}}" id="{{radioSelector.id}}--{{option.index}}" value="{{option.value}}" {% if option.value == radioSelector.value %}checked{% endif %}>{{option.label}}</label>{% endfor %}`,
```
Comment (Snugug) [inline tests/fixtures/input-plugins/input-plugin-radio.js]: If there are multiple radios and only one can be selected, it needs to be w/in a [radiogroup](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/radiogroup)

**Comment by Snugug** [comment]
@scottnath Your instructions are for supporting different major versions, which is not what we're doing here.

I'm not sure we should be relying on the ability to install from a specific branch and base work on that; it's kinda dangerous to do so when we've actually got a live project. If we are going to include a version, it should be `0.0.0` as it will not be updated when the version updates and it'd only be there to allow for the install, not for publishing

**Comment by scottnath** [comment]
@Snugug using version `0.0.0` seems like a great solution.

For the use case...there is sometimes development code which needs to be tested out with other code or used to move forward with other stories before it's PR is complete. In this case, the tests were needed for multiple in-development input plugins. And while this PR is under debate, none of the tests which are needed by @rachelnicole or @ayush2k have received comments or seem to be under debate.

**Comment by scottnath** [comment]
@Snugug looking at other comments now.

File: tests/fixtures/input-plugins/input-plugin-checkbox.js (hunk 3 above)
Comment (scottnath) [inline tests/fixtures/input-plugins/input-plugin-checkbox.js]: @Snugug I think we'll end up deep in weeds with this and we should avoid adding caveats for checking if there are checkboxes and if so, if there are multiple checkboxes. That would be straightforward for a simple use case like this, but I expect any further multi-checkbox plugins will have vastly different structure. 
I'm going to add a fieldset around these checkboxes in this fixture, but not add new functionality to sniff for multiple checkboxes to the [renderer](https://github.com/punchcard-cms/content-types/blob/5550f489f113af109d18b159f3dcb7102b79c830/lib/form/html.js#L113).

Comment (Snugug) [inline tests/fixtures/input-plugins/input-plugin-checkbox.js]: We have very clear semantics on this. If `type === 'checkbox' && options.length > 1`, it needs to be wrapped in a fieldset

Comment (scottnath) [inline tests/fixtures/input-plugins/input-plugin-checkbox.js]: @Snugug 

I understand the _how_ of what you're asking. That is not the issue.

The issue is that this use-case, adding a `fieldset` around a set of checkboxes when there is only one input in an input plugin, will really only apply to _one_ plugin - a plugin which we own. 

I'm sure another developer _may_ create a new input plugin that has only one input that uses checkboxes - but if they did, we should assume the need to create that plugin was beyond what the base input plugin for a single checkbox input could do, _and_ their use-case would call for adding a bunch of other bits of html in the `html` string. Now all those bits will be inside the fieldset too, and those bits may not belong in that fieldset at all.

Also, we would have established that when you create an input plugin that will have multiple checkboxes, there is no need to add a `fieldset` around those checkboxes. So, that would mean that for a developer writing something like [input-plugin-multiple-checkbox](https://github.com/scottnath/content-types/blob/cb470d3592db3c21bc807726dc2d075c49aa9e98/tests/fixtures/input-plugins/input-plugin-multiple-checkbox.js), who wanted to check what the Punchcard creators did with checkboxes, that developer would find a plugin that does not wrap checkboxes and they would (rightfully) assume that the system will wrap fieldsets around all their checkbox sets. Then...to do that in `content-types`, now we will need to search for `{ for option in checkbox.options }` (or any variations on that theme) to add the fieldset around those. 
This is what I meant by saying "I think we'll end up deep in weeds with this"

In short...

* I can easily add this code to `content-types`
* I believe it will set a bad precedent on how to deal with multiple checkboxes from an html standpoint
* I think it would be better to add nunjucks code to deal with adding a fieldset around the checkboxes or not
* fieldsets within fieldsets [look to be legal according to mozilla](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset), so a multi-input plugin for checkboxes would still be html-legal

Comment (scottnath) [inline tests/fixtures/input-plugins/input-plugin-checkbox.js]: @Snugug This is what I've worked up for the fixture:

```
  html: `{% if checkboxSelector.options.length > 1 %}<fieldset id="{{checkboxSelector.id}}"><legend>{{checkboxSelector.label}}</legend>{% endif %}
    {% for option in checkboxSelector.options %}<label for="{{checkboxSelector.id}}--{{option.index}}">
    <input type="{{checkboxSelector.type}}" name="{{checkboxSelector.name}}" id="{{checkboxSelector.id}}--{{option.index}}" value="{{option.value}}" {% if option.value == checkboxSelector.value %}checked{% endif %}>{{option.label}}</label>{% endfor %}
    {% if checkboxSelector.options.length > 1 %}</fieldset>{% endif %}`,
```

Which would then translate to be the code for the actual plugin...if it works for you.

**Comment by scottnath** [comment]
@Snugug changes have been implemented in the PR. The outstanding question is the checkboxes issue.

Also, Node Security is a new check. I see it's non-blocking, but what should we be doing about those responses? Writing module developers?

Please advise or merge.

**Comment by Snugug** [comment]
Node Security tells us when there are security issues identified. If it's with one of our modules, we can go and update, but it's not required as sometimes (like is the case now), the dependency that's out of date and needs to be updated is a deep dependency that we cannot work around until our dependencies are updated. If CI and Coverage pass, and we fail on Node Security with a deep dependency, I'm OK with merging.

**Comment by scottnath** [comment]
@Snugug understood on Node Security

this PR just awaiting your approval of the checkbox solution and merging.
